server/game/core/ongoingEffect/OngoingEffectEngine.ts

```diff
@@ -63,36 +60,64 @@ export class OngoingEffectEngine {
                 handler: () => {
-                    properties.gameAction.setDefaultTarget(() => targets);
-                    if (properties.message && properties.gameAction.hasLegalTarget(context)) {
+                    // TODO Ensure the below line doesn't break anything for a CardTargetSystem delayed effect
+                    properties.immediateEffect.setDefaultTargetFn(() => targets);
+                    if (properties.message && properties.immediateEffect.hasLegalTarget(context)) {
                         let messageArgs = properties.messageArgs || [];
                         if (typeof messageArgs === 'function') {
                             messageArgs = messageArgs(context, targets);
                         }
                         this.game.addMessage(properties.message, ...messageArgs);
                     }
                     const actionEvents = [];
-                    properties.gameAction.queueGenerateEventGameSteps(actionEvents, context);
+                    properties.immediateEffect.queueGenerateEventGameSteps(actionEvents, context);
+                    properties.limit.increment(context.source.owner);
                     this.game.queueSimpleStep(() => this.game.openEventWindow(actionEvents), 'openDelayedActionsWindow');
                     this.game.queueSimpleStep(() => this.game.resolveGameState(true), 'resolveGameState');
                 }
             };
         });
-        if (effectsToRemove.length > 0) {
-            this.unapplyAndRemove((effect) => effectsToRemove.includes(effect));
+        for (const effect of this.effects.filter(
+            (effect) => effect.isEffectActive() && effect.impl.type === EffectName.DelayedEffect
+        )) {
+            const properties = effect.impl.getValue();
+            const triggeringEvents = events.filter((event) => properties.when[event.name]);
+
+            if (triggeringEvents.length > 0) {
+                if (properties.limit.isAtMax(effect.source.owner)) {
+                    effectsToRemove.push(effect);
+                }
+            }
         }
         if (effectTriggers.length > 0) {
-            this.game.openSimultaneousEffectWindow(effectTriggers);
+            // TODO Implement the correct trigger window. We may need a subclass of TriggeredAbilityWindow for multiple simultaneous effects
+            effectTriggers.forEach((trigger) => {
+                trigger.handler();
+            });
+        }
+        if (effectsToRemove.length > 0) {
+            this.unapplyAndRemove((effect) => effectsToRemove.includes(effect));
         }
     }
 
     public removeLastingEffects(card: OngoingEffectSource) {
         this.unapplyAndRemove(
-            (effect) =>
-                effect.matchTarget === card &&
-                effect.duration !== Duration.Persistent
+            (effect) => {
+                if (effect.duration !== Duration.Persistent) {
```

Comment:
I think we might need to do the delayed effect check before this one - since this method is called when a unit leaves the field (or at least should be, currently there's a bug) I believe this check would cause any delayed effect with duration != persistent to be removed when the unit dies. So we should do the delayed effect check first if I'm reading it right.